### PR TITLE
Add a Left Sidebar for Placing Schemas #400

### PR DESCRIPTION
**Pull request by hand-dot**
ref: https://github.com/pdfme/pdfme/issues/400

CleanShot 2024-03-22 at 12 20 41

### CODE REVIEW

**Comment by hand-dot** [comment]
Hi @peteward,
If you have time, Can you check this PR?

File: packages/ui/src/components/Designer/index.tsx [hunk 1]
```diff
@@ -219,59 +226,85 @@ Check this document: https://pdfme.com/docs/custom-schemas`);
...
+        onDragEnd={(event) => {
+          // Triggered after a scheme is dragged & dropped from the left sidebar.
+          if (!event.active) return;
+          const active = event.active;
+
+          const rect = paperRefs.current[pageCursor].getBoundingClientRect();
+          const initialTop = (active.rect.current.initial?.top || 0) - rect.top;
+          const initialLeft = (active.rect.current.initial?.left || 0) - rect.left;
+          const _scale = scale < 1 ? scale + 1 : scale;
+          const adjust = 0.915; // TODO: Investigate later as to why it needs to be adjusted.
```
Comment (hand-dot) [inline packages/ui/src/components/Designer/index.tsx:239]: When dragging an element to the bottom of the page, its position gets misaligned.

Furthermore, if the page is zoomed in, the misalignment becomes even greater.

However, I have determined that this issue does not qualify as a release blocker, and we can fix it in the future.

File: packages/ui/src/components/Designer/index.tsx [hunk 2]
```diff
@@ -27,6 +28,12 @@ import Root from '../Root';
 import ErrorScreen from '../ErrorScreen';
 import CtlBar from '../CtlBar';
 
+const px2mm = (px: number): number => {
```
Comment (peteward) [inline packages/ui/src/components/Designer/index.tsx:31]: we already have px2mm in the image schema, should we add it to the common/helper.ts file where the other conversion functions are?

File: packages/ui/src/components/Designer/index.tsx [hunk 3]
```diff
@@ -219,59 +226,85 @@ Check this document: https://pdfme.com/docs/custom-schemas`);
...
-        activeElements={activeElements}
-        schemas={schemasList[pageCursor] ?? []}
-        changeSchemas={changeSchemas}
-        onSortEnd={onSortEnd}
-        onEdit={(id: string) => {
-          const editingElem = document.getElementById(id);
-          editingElem && onEdit([editingElem]);
+      <DndContext
+        onDragEnd={(event) => {
+          // Triggered after a scheme is dragged & dropped from the left sidebar.
```
Comment (peteward) [inline packages/ui/src/components/Designer/index.tsx:231]:
```suggestion
          // Triggered after a schema is dragged & dropped from the left sidebar.
```

File: playground/src/plugins/signature.ts [hunk 4]
```diff
@@ -57,9 +57,9 @@ export const signature: Plugin<Signature> = {
   pdf: image.pdf,
   propPanel: {
     schema: {},
-
     defaultSchema: {
       type: 'signature',
+      // icon: '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-pen"><path d="M17 3a2.85 2.83 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5Z"/></svg>',
```
Comment (peteward) [inline playground/src/plugins/signature.ts:62]: do you intend to leave this commented?

**Review by peteward — COMMENTED** [review]
This is very cool!

I would be happy to look at the positioning issues soon if you like?

File: playground/src/plugins/signature.ts (hunk 4 above)
Comment (hand-dot) [reply playground/src/plugins/signature.ts:62]: Yes, this is example if not set icon, left sidebar icon will be plugin's Key text

File: packages/ui/src/components/Designer/index.tsx (hunk 2 above)
Comment (hand-dot) [reply packages/ui/src/components/Designer/index.tsx:31]: Ah, yes. It should be.

**Comment by hand-dot** [comment]
Thank you @peteward !

> This is very cool!
> I would be happy to look at the positioning issues soon if you like?

Oh god. Please look this problem.

**Comment by peteward** [comment]
Sure, let's get this merged into `v4` and I will try to help with the positioning. 👍

**Pull request merged**